### PR TITLE
Refactor listener related interface and data structure

### PR DESCRIPTION
**Pull request by critical27**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [X] enhancement

#### Description:
This is a prepare for later poc, we will add new type of listener

1. Reorder data structure, see `ListenersMap` related change.
2. Separate add/remove of space/part for storage and listener.

> BTW, listener related code in community version is out of sync with enterprise version (for example, the base class `Listener`). Why?! Hope we don't have too much conflicts later in enterprise version. I think this PR could unify all usage in both version. @panda-sheep 

## Checklist:
Tests:
- [X] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/common/meta/Common.h [hunk 1]
```diff
@@ -31,29 +31,30 @@ struct PartHosts {
...
   // peers is the part peers which would send logs to the listener
   std::vector<HostAddr> peers_;
 };
 
 using PartsMap = std::unordered_map<GraphSpaceID, std::unordered_map<PartitionID, PartHosts>>;
 // ListenersMap is used for listener replica to get its peers of data replica
 using ListenersMap =
-    std::unordered_map<GraphSpaceID, std::unordered_map<PartitionID, std::vector<ListenerHosts>>>;
+    std::unordered_map<GraphSpaceID,
+                       std::unordered_map<cpp2::ListenerType, std::vector<ListenerHosts>>>;
```
Comment (panda-sheep) [inline src/common/meta/Common.h:57]: Previously `spaceId, partId` ->`mutile type ListenerHost`
Now s`paceId, type` ->`multi partId ListenerHosts`
Is the purpose of this change to make the search faster?

File: src/kvstore/NebulaStore.h [hunk 2]
```diff
@@ -57,6 +56,7 @@ class NebulaStore : public KVStore, public Handler {
   FRIEND_TEST(NebulaStoreTest, ThreeCopiesCheckpointTest);
   FRIEND_TEST(NebulaStoreTest, RemoveInvalidSpaceTest);
   friend class ListenerBasicTest;
+  friend class TopoListenerTest;
```
Comment (panda-sheep) [inline src/kvstore/NebulaStore.h:59]: where it was used?

**Review by panda-sheep — COMMENTED** [review]
Good job

File: src/kvstore/NebulaStore.h (hunk 2 above)
Comment (critical27) [reply src/kvstore/NebulaStore.h:59]: My bad, I'll remove it. It is a legacy code.

File: src/common/meta/Common.h (hunk 1 above)
Comment (critical27) [reply src/common/meta/Common.h:57]: Yes. Consider when we support multiple type of listener. Once we add a new type of listener by `add listener xxx`, we can't efficiently tell whether we already has a listener of type `xxx`, we need to iterate all partitions. And for each partitions, we need to iterate the vector `ListenerHosts` again.

It is a reason that we can't support multiple listener type in the same process.

**Review by pengweisong — COMMENTED** [review]
Good job

File: src/common/meta/Common.h (hunk 1 above)
Comment (panda-sheep) [reply src/common/meta/Common.h:57]: Good, Got it, thx

**Review by panda-sheep — DISMISSED** [review]

File: src/clients/meta/MetaClient.cpp [hunk 3]
```diff
@@ -1068,38 +1068,45 @@ void MetaClient::listenerDiff(const LocalCache& oldCache, const LocalCache& newC
...
-      // new space is added
-      VLOG(1) << "[Listener] SpaceId " << spaceId << " was added!";
-      listener_->onSpaceAdded(spaceId, true);
-      for (const auto& partEntry : spaceEntry.second) {
-        auto partId = partEntry.first;
-        for (const auto& info : partEntry.second) {
-          VLOG(1) << "[Listener] SpaceId " << spaceId << ", partId " << partId << " was added!";
-          listener_->onListenerAdded(spaceId, partId, info);
+      // add new space
+      for (const auto& typeEntry : spaceEntry.second) {
```
[the remark this answers is not in the document]
Comment (critical27) [reply src/clients/meta/MetaClient.cpp:1072]: 👌, hold on a sec.

[the remark this answers is not in the document]
Comment (critical27) [reply src/clients/meta/MetaClient.cpp]: Done

**Review by pengweisong — APPROVED** [review]

**Review by panda-sheep — APPROVED** [review]

**Pull request merged**